R/input-select.R

```diff
@@ -186,6 +186,17 @@ needOptgroup <- function(choices) {
 #'   value when it is a single choice input and the empty string is not in the
 #'   `choices` argument. This is to keep compatibility with
 #'   `selectInput(..., selectize = FALSE)`.
+#'   
+#'   The selectize input created from `selectInput(..., selectize = TRUE, 
+#'   multiple = TRUE)` allows deletion of all selected options, which will 
+#'   then return NULL as its value. Because the default behavior of an 
+#'   `observeEvent()` is to ignore expressions that evaluate to NULL, a change 
+#'   from a single selected option to no selected options in such a 
+#'   `selectInput()` would not constitute an event that would trigger an 
+#'   `observeEvent()` by default. If this behavior is undesirable in your 
+#'   context, it may be useful to set the `ignoreNULL` (and perhaps also the 
+#'   `ignoreInit`) parameter of `observeEvent()` to TRUE; see [observeEvent()]
+#'   for details. 
```

Comment:
Just a few small edits here.

```suggestion
#'   When `multiple = TRUE` is used with `selectizeInput()`, users can select
#'   multiple options as well as delete all selected options. When no options
#'   are selected, the server will receive a value of `NULL`. Because the default
#'   behavior of an `observeEvent()` is to ignore expressions that evaluate to NULL,
#'   a change from a single selected option to no selected options in such a 
#'   `selectizeInput()` will not invalidate (cause to update) an 
#'   `observeEvent()` by default. If this behavior is undesirable in your 
#'   context, it may be useful to set `ignoreNULL = TRUE` in `observeEvent()`
#'   (and possibly also `ignoreInit = TRUE` as well). See [observeEvent()] for details. 
```

Reply:
This looks great to me, and certainly would have been a help to me in understanding this issue!

Reply:
Awsome, thanks!

Reply:
This probably isn't the place for this, per se, but I have also been trying to figure out how to know instinctively if a req() is required (very meta!) in a use case or not. The help pages don't necessarily make this clear, currently.

But I do seem to have a good example of such a situation. In my current App, I have a selectInput() being created by a renderUI() whose choices depend on the input from another selectInput(). When the App starts up, the "upstream" selectInput() briefly holds a value of NULL, causing the second selectInput() to return an error (thanks to an if check inside the renderUI). I'm thinking this is a good example of where a req() solves an otherwise thorny problem...

REPREX:
```
library(shiny)

ui <- fluidPage(

 uiOutput("input1"),
 uiOutput("input2")
 
)

# Define server logic required to draw a histogram
server <- function(input, output) {

 output$input1 <- renderUI({
   
   selectInput("first_picker", 
               "Pick a first choice",
               choices = c("A", "B", "C"))
   
 })
 
 output$input2 <- renderUI({
   req(input$first_picker) #This req() is required to avoid an error with the if statement below on startup.
   
   if(input$first_picker == "A")
   { new_choices = c(1,2,3)
   } else {
     if(input$first_picker == "B") {
     new_choices = c(4,5,6)
     } else { new_choices = c(7,8,9) }
   }
     selectInput("second_picker", 
               "Pick a second choice",
               choices = new_choices)
 })
  
}

# Run the application 
shinyApp(ui = ui, server = server)
```

Is there a place I could maybe put this? I can't find the code for req() on the repo...Or is there a clever way to sidestep the need for req(), and if so, when is req() actually _required_ to solve a problem?

Reply:
Hmm... I think it's probably useful to think of `observeEvent()` as the weird one in this case, as in most other places you _do_ need to handle the `NULL` value. The most consistent thing would probably be for `observeEvent()` to not ignore `NULL` values, although I'm sure there's some history there. In your case, whether or not to use `req()` is dependent upon whether or not you want to see the second select input if there's not selection in the first. If you don't want to see the second, then use `req()` to make it dependent.

If you do want to see the second input regardless of the value of `first_input`, then you'll need to handle the `NULL` value yourself. Here's a version that does this:

```r
library(shiny)

ui <- fluidPage(
  selectInput("first_picker", "Pick a first choice", choices = c("A", "B", "C")),
  selectInput("second_picker", "Pick a second choice", choices = NULL)
)

server <- function(input, output, session) {
  second_choices <- reactive({
    first <-
      if (!is.null(input$first_picker)) {
        input$first_picker
      } else {
        "A" # the default value if first_picker is NULL
      }

    switch(
      first,
      "A" = c(1, 2, 3),
      "B" = c(4, 5, 6),
      "C" = c(7, 8, 9)
    )
  })

  observe(updateSelectInput(session, "second_picker", choices = second_choices()))
}

# Run the application
shinyApp(ui = ui, server = server)
```

If it's important to use `uiOuput()`, you can use a similar strategy of creating a reactive that always returns a valid value by accounting for the potential `NULL`.

Hope this helps!